### PR TITLE
Price API

### PR DESCRIPTION
**Pull request by adityapk00**
Stub implementation for the Prices API. Mainly to review the API

### CODE REVIEW

**Comment by adityapk00** [comment]
Updated with full implementation, both for current and historical prices. 

- Historical prices are cached and stored locally
- Current price is fetched from 3 sources, outliers discarded, then the median is used

**Comment by gmale** [comment]
Accidentally closed inside zenhub. Fixed.

**Comment by gmale** [comment]
Update: in the next sprint, @LarryRuane would like to add a few tests, complete a review and merge this work.

**Comment by nighthawk24** [comment]
Any update on this? @LarryRuane 
@adityapk00 do you have some bandwidth to resolve the merge conflicts?

**Comment by LarryRuane** [comment]
Hi @nighthawk24, thanks for the reminder, sorry for the delay. I rebased this branch onto latest master (also resolving the merge conflicts), and pushed it to a new branch of my fork: https://github.com/LarryRuane/lightwalletd/tree/adityapk00-prices_stub

Aditya, if that looks okay, I can force-push those commits onto your branch here (if I have permission, which I think I do). I don't like to force-push onto someone else's branch without asking first.

I did some basic testing (on my branch), and your changes seem to work. If you and Adi are happy with it, I'd like to do just a little more review and then I'll push to master. I really don't think we need to spend a lot of effort adding tests, because this is kind of an add-on, not part of the critical core of lightwalletd. 

One problem I ran into when rebasing onto latest master is that running `gofmt` on the new file, `common/prices.go`, (which my editor automatically does) had the effect of changing every line. This is because every line ended with a control-M character. So this file had a DOS format, instead of Unix. We really need all code (except the stuff in `vendor`) to stay up-to-date formatted with `gofmt`; I hope this doesn't cause you any trouble. (If you diff with the ignore-whitespace in effect, such as `git diff -b` or similar, you'll see that there are no code changes.)

**Comment by adityapk00** [comment]
Looks good to me. Sorry about the whitespace thing - I use a windows machine and I think it does that by default. No problem from my side to reformat

File: common/prices.go
Comment (defuse) [inline common/prices.go:123]: There are two ways I'd like to see this function improved:

1. It's possible for this function to return -1 without an error. This happens if all of the API endpoints fail. It turns out that some code higher up the call stack will catch this, but this is error-prone because newly added APIs might fail to make the same check. It would be clearer if this code returned an error in that case rather than a value of -1 without an error. The code in the higher-level API function should then just check for the error condition.
2. I would like to have the property that at least two providers must agree on the price (within 20%, say). This is currently true of this function as-sis, but it relies on there being at least three providers in the list. If one is deleted, then the code will return that provider's value without it being cross-checked by another one. I suggest changing the code below so that at least to providers must agree.

**Review by defuse — COMMENTED** [review]
Adding partial review comments. I have yet to review the historical price functionality and the higher-up-the-call-stack code.

**Comment by nighthawk24** [comment]
Bump. Please consider including this in the next major release.

**Comment by LarryRuane** [comment]
Force-pushed rebase to latest master branch, fixed a bug in the average price calculation, and also disabled this when running in darkside wallet test mode. I still need to address @defuse's comment, and see if I can clean up the code a bit. But this should be close to being able to merge.

File: walletrpc/service.proto
Comment (ccjernigan) [inline walletrpc/service.proto:140]: Is the request a single item or a list?  The comment and the field declaration look like they might be inconsistent.

File: common/prices.go
Comment (LarryRuane) [reply common/prices.go]: The latest commit implements both of these suggestions (thanks for them!). I manually tested the second suggestion by deleting one of the 3 providers, and since the remaining two are close, the result is their average (which is shown in the third line):
```
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":114.2,"provider":"github.com/zcash/lightwalletd/common.fetchBinancePrice","time":"2022-05-17T19:14:35-06:00"}
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":114.67531021588819,"provider":"github.com/zcash/lightwalletd/common.fetchCoinCapPrice","time":"2022-05-17T19:14:35-06:00"}
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":114.4376551079441,"time":"2022-05-17T19:14:35-06:00"}
```
Then I added 100 to one of the provider's result (still only two providers), and the result is as desired:
```
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":114.2,"provider":"github.com/zcash/lightwalletd/common.fetchBinancePrice","time":"2022-05-17T19:15:39-06:00"}
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":214.7104225164866,"provider":"github.com/zcash/lightwalletd/common.fetchCoinCapPrice","time":"2022-05-17T19:15:39-06:00"}
{"app":"lightwalletd","error":"Discarding price (114.20) because too far away from median (164.46","level":"error","method":"CurrentPrice","msg":"Service","time":"2022-05-17T19:15:39-06:00"}
{"app":"lightwalletd","error":"Discarding price (214.71) because too far away from median (164.46","level":"error","method":"CurrentPrice","msg":"Service","time":"2022-05-17T19:15:39-06:00"}
{"app":"lightwalletd","level":"error","msg":"Error getting prices from web APIs: insufficient price providers are available","time":"2022-05-17T19:15:39-06:00"}
```

**Comment by LarryRuane** [comment]
Force-pushed to fix some things (see commit message), I think this is very close to being able to merge, if the API is acceptable (@adityapk00 @nighthawk24 @pacu @ccjernigan). I would appreciate also if @defuse could take another look. I added @steven-ecc but don't feel any obligation.

The only things I'm aware of that are needed before merging are:
- undo some test-only code that will make this easier for reviewers to test and experiment with (search for "XXX testing") 
- remove some of the logging (but it's nice for manual testing)

File: common/prices.go
Comment (ccjernigan) [reply common/prices.go]: I'm new to the changes and all the implementation here, but if we're relying on multiple providers does that mean we're subject to denial of service if even one of those services goes down temporarily?

**Comment by LarryRuane** [comment]
Force-pushed (twice, [first](https://github.com/zcash/lightwalletd/compare/65fda0207483f9bc0f53a57b0527447c820b6e31..baf3e7c9cfb0ea3ba368951f19842404b9978801), [second](https://github.com/zcash/lightwalletd/compare/baf3e7c9cfb0ea3ba368951f19842404b9978801..a7bd7741cf1d96efd26b69400160d3c3c2e978ff)) to make these changes:
- A call to `GetCurrentPrice` gRPC soon after startup returns a confusing error, "price too old"; instead return "starting up, prices not available yet"
- Don't hold the mutex during price fetching. This allows `GetCurrentPrice` to return immediately if our price oracles are slow to respond.

I also did some manual testing: I turned off my internet, and the price fetches timed out within about 10 seconds. But during this time, the currrent price (which is cached every 15 minutes) can still be fetched using `GetCurrentPrice`. (A historical price could be fetched if it had been fetched and cached before.)

File: common/prices.go
Comment (LarryRuane) [reply common/prices.go]: I verified that our requests to the providers do timeout after about 10 seconds, but during that time, gRPC requests are still processed (other than, of course, requests to `coingecko` for historical data, but in that case, only `GetZECPrice` gRPCs are delayed, not other gRPCs, this server is multithreaded). So a non-responding provider would not cause DoS to our (wallet) clients.

**Comment by LarryRuane** [comment]
In case this is helpful to reviewers, you can test without an actual wallet like this:
```
$ grpcurl -plaintext  localhost:9067 cash.z.wallet.sdk.rpc.CompactTxStreamer/GetCurrentZECPrice
{
  "timestamp": "1652900224",
  "currency": "USD",
  "price": 102.69499999999998
}
$ grpcurl -plaintext -d '{"timestamp":1652895450, "currency":"USD"}' localhost:9067 cash.z.wallet.sdk.rpc.CompactTxStreamer/GetZECPrice
{
  "timestamp": "1652873850",
  "currency": "USD",
  "price": 114.12326794636196
}
$
```
You can obtain the current "Unix" time (seconds since 1970) this way:
```
$ date '+%s'
1652901832
$ 
```
You can find the Unix time for a given date like this (the date string is mostly free-format human readable):
```
$ date '+%s' --date='jan 15 2020'
1579071600
$
```

**Comment by LarryRuane** [comment]
[Force-pushed](https://github.com/zcash/lightwalletd/compare/a7bd7741cf1d96efd26b69400160d3c3c2e978ff..1a7ebe32914b3c08ffb441d33ae6aa073fcf7af6) again to remove test code. If you want to restore the test code for your own testing, run this patch command on the latest commit:

<details>
<summary>Click to expand patch command</summary>

```
patch -R -p1 <<EOF
--- a/common/prices.go
+++ b/common/prices.go
@@ -235,19 +235,14 @@ func GetCurrentPrice() (float64, error) {
 
 // return the time in YYYY-MM-DD string format
 func day(t time.Time) string {
-	// XXX testing: pretend every second is a new day:
-	return t.Format("2006-01-02-03-04-05")
-
-	// XXX testing: before merge, restore this:
-	//return t.Format("2006-01-02")
+	return t.Format("2006-01-02")
 }
 
 // GetHistoricalPrice returns the price for the given day, but only
 // accurate to day granularity.
 func GetHistoricalPrice(ts time.Time) (float64, *time.Time, error) {
 	dt := day(ts)
-	// XXX testing (change back to "2006-01-02" before merging)
-	canonicalTime, err := time.Parse("2006-01-02-03-04-05", dt)
+	canonicalTime, err := time.Parse("2006-01-02", dt)
 	if err != nil {
 		return -1, nil, err
 	}
@@ -344,12 +339,8 @@ func StartPriceFetcher(dbPath string, chainName string) {
 				latestPriceTime = time.Now()
 				mutex.Unlock()
 			}
-
-			// XXX testing: run each minute
-			time.Sleep(1 * time.Minute)
-
-			// XXX testing: before merge, restore this:
-			//time.Sleep(15 * time.Minute)
+			// price data 15 minutes out of date is probably okay
+			time.Sleep(15 * time.Minute)
 		}
 	}()
 }
EOF
```
</details>

**Comment by LarryRuane** [comment]
Force-pushed to rebase onto master, ping @steven-ecc @ccjernigan @defuse for review, please note that the following files are generated and don't need review:
- `docs/rtd/index.html`
- `walletrpc/service.pb.go`
- `walletrpc/service_grpc.pb.go`

File: cmd/root.go
Comment (LarryRuane) [reply cmd/root.go:276]: Hmm, now that you mention it, I may have added that conditional in haste, incorrectly. I was thinking well _of course_ we don't want lightwalletd to go out and ping these websites in dark mode, since it's supposed to be self-contained (no `zcashd`) testing only. But maybe it's okay to do that.

OR, we could stub out the two price gRPCs and just return hardcoded fake data. That way internet access wouldn't be needed to do (full) darkside testing. Which do you think would be better?

Comment (pacu) [reply cmd/root.go:276]: we should return stub data so integration tests can be reliable and deterministic

**Comment by LarryRuane** [comment]
Rebased onto latest master, and add commit to support darksidewallet testing - return fixed price data in darkside mode. I think this should be ready to merge now (pending review).

File: common/prices.go
Comment (daira) [inline common/prices.go:157]: Requiring at least two sources means that either of them can determine the price (since we have to assume they know the price from the other source, then they can choose whatever median they want).

I think at least three sources should be required. In that case no single source can choose an outlier price: a malicious/corrupted source can only cause the median to be any price between the other two.

**Comment by LarryRuane** [comment]
Force-pushed to address @daira's suggestion, now require at least 3 oracles. I also added a fourth oracle, `coingecko` (which we're already using for historical data, but it can also provide current data), because otherwise all oracles would need to be available. This way, one of them can be unreachable and we can still establish a current price.

**Review by pacu — APPROVED** [review]
utACK. Left some minor recommendation regarding an error message.

**Review by pacu — APPROVED** [review]

**Comment by LarryRuane** [comment]
Force-pushed rebased

File: walletrpc/service.proto
Comment (pacu) [inline walletrpc/service.proto:142]: What Timestamp should one send to get the latest available?

**Review by pacu — APPROVED** [review]
LGTM! 

Just a doubt on API usage

File: walletrpc/service.proto
Comment (LarryRuane) [reply walletrpc/service.proto:142]: If you want the latest, don't use `GetZECPrice` (which takes this timestamp argument); that's intended only for historical data. Instead, it's recommended to use `GetCurrentZECPrice`.

I suppose you could use `GetZECPrice` and specify the current time, but then you're getting the price from only one oracle; `GetCurrentZECPrice` uses up to four oracles, so it's likely to be more accurate.

**Comment by defuse** [comment]
The binance URL will need to be updated to `ZECBUSD`, see: https://github.com/adityapk00/lightwalletd/issues/12

**Comment by LarryRuane** [comment]
Closing, I don't think we're going to be doing this, please reopen if needed, thanks.

**Comment by str4d** [comment]
Indeed we don't want to do this; there is no authentication of the price data, so it would enable lightwalletd operators to manipulate the exchange rates being served to light clients.

The `zcash_client_backend 0.13` Rust crate adds APIs for doing effectively the same querying as this PR, but light clients themselves query the exchanges so they get authenticated exchange rate data (and do so over Tor for IP hiding).

**Pull request closed without merging**
